File: src/main/java/dev/psygamer/wireframe/util/math/HitResult.java
package dev.psygamer.wireframe.util.math;

import dev.psygamer.wireframe.util.math.vector.Vector3d;

public abstract class HitResult {
	
	public enum Type {
		MISS, BLOCK, ENTITY
	}
	
	protected final Vector3d location;
	
	protected HitResult(final Vector3d location) {
		this.location = location;
	}
	
	public double distanceTo(final Vector3d other) {
		return Math.sqrt(squaredDistanceTo(other));
	}
	
	public double squaredDistanceTo(final Vector3d other) {
		final double xDiff = this.location.getX() - other.getX();
		final double yDiff = this.location.getY() - other.getY();
		final double zDiff = this.location.getZ() - other.getZ();
		
		return xDiff * xDiff + yDiff * yDiff + zDiff * zDiff;
	}
	
	public Vector3d getLocation() {
		return this.location;
	}
}

File: src/main/kotlin/dev/psygamer/wireframe/util/math/HitResult.kt
package dev.psygamer.wireframe.util.math

import org.joml.Vector3d
import kotlin.math.sqrt

abstract class HitResult protected constructor(val location: Vector3d) {
	enum class Type {
		MISS, BLOCK, ENTITY
	}

	fun distanceTo(other: Vector3d): Double {
		return sqrt(squaredDistanceTo(other))
	}

	fun squaredDistanceTo(other: Vector3d): Double {
		val xDiff = location.x - other.x
		val yDiff = location.y - other.y
		val zDiff = location.z - other.z
		return xDiff * xDiff + yDiff * yDiff + zDiff * zDiff
	}
}